### PR TITLE
Port Classic Flang to LLVM 14

### PR DESCRIPTION
**Pull request by bryanpkc**
This PR ports Classic Flang-related changes from [flang-13.0.0-20220325](https://github.com/flang-compiler/classic-flang-llvm-project/commit/0254365051067b50372914150a004048c5cbce54) to [llvmorg-14.0.1](https://github.com/flang-compiler/classic-flang-llvm-project/commit/c62053979489ccb002efe411c3af059addcb5d7d), on which our new `release_14x` branch will be based.

I have squashed some commits together to resolve merge conflicts and make them buildable; in such cases, the original commits are cited. In some cases, I have undone the squashes performed by @abrahamtovarmob and separated commits that have clearly different purposes. Some upstream commits which were backported to `release_13x` did not need to be ported to this branch (89065d4abfe506af1da432ddf8651f1893a4a4fd, b4723b90ebe016fd0973c4afeffd97f0d7c3b66a, 0254365051067b50372914150a004048c5cbce54).

A commit was added on top of the branch to enable GitHub builds of the `release_14x` branch.

### CODE REVIEW

File: clang/include/clang/Driver/Options.td
Comment (RichBarton-Arm) [inline clang/include/clang/Driver/Options.td]: I think this comment is supposed to live with the declaration of gfortran_Group? Doesn't seem to match the intention of these classic flang option groups?

File: clang/include/clang/Driver/Options.td
Comment (RichBarton-Arm) [inline clang/include/clang/Driver/Options.td]: frontend_optimize is in the wrong group I think. Can stay in the gfortran_Group

File: clang/include/clang/Driver/Options.td
Comment (RichBarton-Arm) [inline clang/include/clang/Driver/Options.td]: I'm concerned about using ENABLE_CLASSIC_FLANG to guard additions to Options.td when it does not also guard alterations to Options.td. That seems confusing to me. I don't think there is any chance of an unintended downside to having classic flang options defined in the compiled opt table when classic flang is not included? They simply won't be used by the rest of the code.

File: clang/include/clang/Driver/Options.td
Comment (RichBarton-Arm) [inline clang/include/clang/Driver/Options.td:4726]: Although this is quite nice and clear, despite my comment above about not liking ENABLE_CLASSIC_FLANG use like this....

Perhaps I could be persuaded, but IMO we should just modify this file for classic flang without using ENABLE_CLASSIC_FLANG. What do others think?

File: clang/lib/Driver/Driver.cpp
Comment (RichBarton-Arm) [inline clang/lib/Driver/Driver.cpp:4655]: Need to pull in https://github.com/flang-compiler/classic-flang-llvm-project/pull/100/files

**Review by RichBarton-Arm — COMMENTED** [review]
I have recently done this exercise downstream so I took a look. I left a few comments in passing for minor things.

It looks like you need to cherry pick across my three recent PR #100 , #102  and #104 .

When doing this downstream, I found a regression in the way classic flang handles preprocessing with `-save-temps` (and some other not important options). `flang -E -save-temps file.F95` causes the preprocessor to be run twice. I have a fix so perhaps it is easiest if I make a separate PR and we can commit it over the top of this one?

**Comment by bryanpkc** [comment]
I added a patch to revert an ancient change in LoopVectorize.cpp, which was causing an AArch64 test to fail. This probably needs a review by @vjayathirtha-nv or someone more familiar with the motivation of the original patch.

**Comment by bryanpkc** [comment]
@RichBarton-Arm  Thanks for the review; I'll address your concerns soon.

**Comment by bryanpkc** [comment]
> When doing this downstream, I found a regression in the way classic flang handles preprocessing with `-save-temps` (and some other not important options). `flang -E -save-temps file.F95` causes the preprocessor to be run twice. I have a fix so perhaps it is easiest if I make a separate PR and we can commit it over the top of this one?

That sounds like a good idea. In this PR I was only porting the basic patches from `flang-13.0.0-20220325` so there would have been a follow-up PR to cherry-pick other more recent patches. I also didn't notice the double preprocessing problem in our downstream compiler previously.

File: clang/include/clang/Driver/Options.td
Comment (bryanpkc) [reply clang/include/clang/Driver/Options.td]: I agree with you. Simply porting the old inconsistent guards to LLVM 14 is easy, but we really ought to make these things consistent. I will take a deeper look at the uses of `ENABLE_CLASSIC_FLANG` in this file. This is a good opportunity to make the file compatible with both LLVM Flang and Classic Flang (we are playing with both downstream).

**Comment by vjayathirtha-nv** [comment]
> I added a patch to revert an ancient change in LoopVectorize.cpp, which was causing an AArch64 test to fail. This probably needs a review by @vjayathirtha-nv or someone more familiar with the motivation of the original patch.

Hi @bryanpkc, [this comment](https://github.com/flang-compiler/llvm/commit/68669092a5ad1fd29f62346dc30973f86cfacbfe#commitcomment-32754393) describes the context for this change. However I'm not in charge of maintaining classic flang anymore, maybe @sscalpone or @schweitzpgi can help you further.

File: clang/lib/Driver/ToolChains/ClassicFlang.cpp
Comment (guy-david) [inline clang/lib/Driver/ToolChains/ClassicFlang.cpp:358]: Clang has moved to DWARF 5 by default, should this be updated as well?

Comment (bryanpkc) [reply clang/lib/Driver/ToolChains/ClassicFlang.cpp:358]: The option `-Mx,120,0x2000000` should enable DWARF 5, but I am not sure how well it is supported; more thorough testing and perhaps some additional fixes might be needed. I will create an issue to track this work.

File: clang/lib/Driver/Driver.cpp
Comment (bryanpkc) [reply clang/lib/Driver/Driver.cpp:4655]: Done.

File: clang/include/clang/Driver/Options.td
Comment (bryanpkc) [reply clang/include/clang/Driver/Options.td]: I have updated the uses of the macro in the file so that the conflicting Fortran options are better guarded.

File: clang/include/clang/Driver/Options.td
Comment (bryanpkc) [reply clang/include/clang/Driver/Options.td]: Fixed.

File: clang/include/clang/Driver/Options.td
Comment (bryanpkc) [reply clang/include/clang/Driver/Options.td]: The comment has been rewritten, and the option groups have been moved to be closer to the group member options.

**Comment by bryanpkc** [comment]
flang-compiler/flang#1242 will fix the test failures seen in the "Build and test Flang" jobs in this PR.

File: clang/include/clang/Driver/Options.td
Comment (RichBarton-Arm) [inline clang/include/clang/Driver/Options.td]: Seems like an unnecessary bit of delta in classic flang that could be removed?

File: clang/include/clang/Driver/Options.td
Comment (RichBarton-Arm) [inline clang/include/clang/Driver/Options.td:4676]: Unnecessary delta

File: clang/include/clang/Driver/Options.td
Comment (RichBarton-Arm) [inline clang/include/clang/Driver/Options.td:4683]: unnecessary delta

File: clang/include/clang/Driver/Options.td
Comment (RichBarton-Arm) [reply clang/include/clang/Driver/Options.td:4726]: For the record, I am very happy with the new approach that lets classic flang sit next to llvm flang wrt Options.td.
Sterling work @bryanpkc

File: clang/include/clang/Driver/ToolChain.h
Comment (RichBarton-Arm) [inline clang/include/clang/Driver/ToolChain.h:605]: Needs the ENABLE_CLASSIC_FLANG guard?

File: clang/include/clang/Driver/ToolChain.h
Comment (RichBarton-Arm) [inline clang/include/clang/Driver/ToolChain.h:707]: Needs the ENABLE_CLASSIC_FLANG guard?

File: clang/lib/CodeGen/CGDebugInfo.cpp
Comment (RichBarton-Arm) [inline clang/lib/CodeGen/CGDebugInfo.cpp:3889]: Changes in file need the ENABLE_CLASSIC_FLANG guard?

File: clang/lib/Driver/CMakeLists.txt
Comment (RichBarton-Arm) [inline clang/lib/Driver/CMakeLists.txt]: I wonder if that insertion might be more obvious inserted at the same line as the deletion of Flang.cpp. Minor nit though.

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (RichBarton-Arm) [inline clang/lib/Driver/ToolChains/Clang.cpp:4897]: I think this whole block also needs an include guard as it uses `OPT_Minfo_EQ` and friends? So I think this branch currently wouldn't build with `-DLLVM_ENABLE_CLASSIC_FLANG=Off`

Comment (RichBarton-Arm) [reply clang/lib/Driver/ToolChains/Clang.cpp:4897]: Weird also how this block needs to be in the Clang driver and can't just be in ClassicFlang.cpp

File: clang/lib/Driver/ToolChains/ClassicFlang.cpp
Comment (RichBarton-Arm) [reply clang/lib/Driver/ToolChains/ClassicFlang.cpp:358]: I'd be very afraid to change default here. Better the devil you know.

File: clang/lib/Driver/ToolChains/CommonArgs.cpp
Comment (RichBarton-Arm) [inline clang/lib/Driver/ToolChains/CommonArgs.cpp:127]: Needs preprocessor guard?

**Review by RichBarton-Arm — COMMENTED** [review]
Looking great Bryan! I like the solution to Options.td preprocessing now, thanks for taking time to do that. 

I notice that there are a number of places in the code where the classic flang changes are not guarded with `ENABLE_CLASSIC_FLANG` Some of those are harmless, but it looks to my eye like some of them need guarding after your changes to Options.td preprocessing. Worth a look before we approve.

File: clang/include/clang/Driver/Options.td
Comment (bryanpkc) [reply clang/include/clang/Driver/Options.td]: Actually I think this change should be upstreamed, but `'fortran'` should be replaced with `'flang'`.

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (bryanpkc) [reply clang/lib/Driver/ToolChains/Clang.cpp:4897]: I have guarded the whole block with `ENABLE_CLASSIC_FLANG`. I will create a follow-up issue to investigate moving this code to a more appropriate location.

File: clang/lib/Driver/ToolChains/CommonArgs.cpp
Comment (bryanpkc) [reply clang/lib/Driver/ToolChains/CommonArgs.cpp:127]: Fixed.

File: clang/lib/Driver/CMakeLists.txt
Comment (bryanpkc) [reply clang/lib/Driver/CMakeLists.txt]: Fixed.

File: clang/include/clang/Driver/ToolChain.h
Comment (bryanpkc) [reply clang/include/clang/Driver/ToolChain.h:707]: Fixed.

File: clang/include/clang/Driver/ToolChain.h
Comment (bryanpkc) [reply clang/include/clang/Driver/ToolChain.h:605]: Fixed.

**Comment by bryanpkc** [comment]
flang-compiler/flang#1244 is needed to fix several AArch64 Flang test cases.

File: clang/lib/CodeGen/CGDebugInfo.cpp
Comment (bryanpkc) [reply clang/lib/CodeGen/CGDebugInfo.cpp:3889]: Unlike the changes we made in Driver, the optional `flags` field we added to the `DIGlobalVariable` metadata has not caused much problem or confusion historically. Inserting `#ifdef` in all the places where we use the debuginfo metadata APIs will likely create much more verbose delta, but no improvement in clarity.

**Comment by bryanpkc** [comment]
@RichBarton-Arm  Could you take a look at this again?

File: clang/lib/Driver/ToolChains/ClassicFlang.h
Comment (RichBarton-Arm) [inline clang/lib/Driver/ToolChains/ClassicFlang.h]: Shouldn't this be ..._ClassicFlang_H ?

**Review by RichBarton-Arm — APPROVED** [review]
Two comments, but from the clang/driver side I think this merge is good to go with those two addressed. It would be good for someone to also look over the Debug changes here to check they are all legit.

My two comments:
 1. It looks like this commit is not present in this port, but I expected it to be: https://github.com/flang-compiler/classic-flang-llvm-project/commit/89065d4abfe506af1da432ddf8651f1893a4a4fd @pawosm-arm  can confirm if that's right.
 2. I left one inline comment also

File: clang/lib/Driver/ToolChains/ClassicFlang.h
Comment (pawosm-arm) [reply clang/lib/Driver/ToolChains/ClassicFlang.h]: The commit mentioned by Richard should be in this port.

Comment (bryanpkc) [reply clang/lib/Driver/ToolChains/ClassicFlang.h]: The commit is not in this port because it is already part of the baseline llvmorg-14.0.1 (commit 26f56438e3dab44cea4c8f16d4cb16e9424b02c6).

Comment (bryanpkc) [reply clang/lib/Driver/ToolChains/ClassicFlang.h]: The macro has been fixed.

**Review by pawosm-arm — APPROVED** [review]

**Comment by kiranchandramohan** [comment]
ping @shivaramaarao

**Review by shivaramaarao — APPROVED** [review]

**Pull request merged**
